File: tests/leap_compat/native/NativeProbe.kt
package nativeprobe

import ai.liquid.leap.HiddenStates
import uniffi.cera_ffi.BackendPreference
import uniffi.cera_ffi.CeraEngine
import uniffi.cera_ffi.EngineConfig
import uniffi.cera_ffi.FfiException
import uniffi.cera_ffi.Session
import uniffi.cera_ffi.SessionConfig
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.file.Files
import java.nio.file.Path

// Raw CPU boundary experiment. This is not a ModelRunner implementation.
class CeraNativeProbe(
    val session: Session,
) {
    fun hiddenStates(tokens: List<UInt>): HiddenStates {
        val bytes = session.hiddenStatesForTokens(tokens)
        val dimensions = session.hiddenSize().toInt()
        check(bytes.size.toLong() == tokens.size.toLong() * dimensions * 4)
        val data = FloatArray(bytes.size / 4)
        ByteBuffer
            .wrap(bytes)
            .order(ByteOrder.LITTLE_ENDIAN)
            .asFloatBuffer()
            .get(data)
        return HiddenStates(data, tokens.size, dimensions)
    }
}

fun main(args: Array<String>) {
    val config = EngineConfig(contextSize = 32uL, backend = BackendPreference.CPU)
    try {
        CeraEngine.fromBytes(byteArrayOf(0, 1, 2), config).close()
        error("Invalid model bytes were accepted")
    } catch (_: FfiException) {
    }
    val engine = CeraEngine.fromBytes(Files.readAllBytes(Path.of(args[0])), config)
    val tokens = engine.encodeText("ab")
    check(tokens == listOf(0u, 1u))
    val configForSession = SessionConfig(maxSeqLen = 32u, seed = 7uL)
    val probe = CeraNativeProbe(engine.newSession(configForSession))
    val control = engine.newSession(configForSession)
    val options = engine.defaultGenerateOpts().copy(maxTokens = 2u, temperature = 0f)
    engine.close()
    try {
        probe.session.appendTokens(tokens)
        control.appendTokens(tokens)
        val position = probe.session.position()
        // Different content and length detect a destructive reset-and-replay of live KV.
        val queryTokens = listOf(1u, 0u, 1u)
        val states = probe.hiddenStates(queryTokens)
        check(states.tokenCount == 3 && states.embeddingDim == 32)
        check(probe.session.position() == position)
        try {
            probe.hiddenStates(listOf(9999u))
            error("Invalid token was accepted")
        } catch (_: FfiException.InvalidToken) {
        }
        check(probe.session.position() == position)
        val copied = states[0]
        val original = states.data[0]
        copied[0] = original + 1f
        check(states.data[0] == original)
        val actual = probe.session.generate(options).tokens
        val expected = control.generate(options).tokens
        check(actual == expected && actual.size == 2)
        check(states.data.all { it.isFinite() })
        val bits = states.data.map { it.toRawBits().toUInt() }
        println("{\"tokens\":$tokens,\"query_tokens\":$queryTokens,\"hidden_bits\":$bits,\"generated\":$actual,\"position\":$position}")
    } finally {
        probe.session.close()
        control.close()
    }
}

File: tests/leap_compat/native/NativeProbe.swift
import Cera
import Foundation
import LeapSDK

// Raw CPU boundary experiment. This is not a ModelRunner implementation.
struct CeraNativeProbe {
  let session: Cera.Session

  func hiddenStates(tokens: [UInt32]) throws -> LeapSDK.HiddenStates {
    let bytes = try session.hiddenStatesForTokens(tokens: tokens)
    let dimensions = Int(session.hiddenSize())
    precondition(bytes.count == tokens.count * dimensions * 4)
    let data = KotlinFloatArray(size: Int32(bytes.count / 4))
    bytes.withUnsafeBytes { raw in
      for index in 0..<(bytes.count / 4) {
        let word = UInt32(
          littleEndian: raw.loadUnaligned(fromByteOffset: index * 4, as: UInt32.self))
        data.set(index: Int32(index), value: Float(bitPattern: word))
      }
    }
    return LeapSDK.HiddenStates(
      data: data, tokenCount: Int32(tokens.count), embeddingDim: Int32(dimensions))
  }
}

@main
struct NativeProbeMain {
  static func main() throws {
    let config = Cera.EngineConfig(contextSize: 32, backend: .cpu)
    do {
      _ = try Cera.CeraEngine.fromBytes(bytes: Data([0, 1, 2]), config: config)
      fatalError("Invalid model bytes were accepted")
    } catch is Cera.FfiError {}

    let bytes = try Data(contentsOf: URL(fileURLWithPath: CommandLine.arguments[1]))
    var engine: Cera.CeraEngine? = try Cera.CeraEngine.fromBytes(bytes: bytes, config: config)
    weak let weakEngine = engine
    let tokens = engine!.encodeText(text: "ab")
    precondition(tokens == [0, 1])
    let configForSession = Cera.SessionConfig(maxSeqLen: 32, seed: 7)
    let probe = CeraNativeProbe(session: try engine!.newSession(config: configForSession))
    let control = try engine!.newSession(config: configForSession)
    var options = engine!.defaultGenerateOpts()
    options.maxTokens = 2
    options.temperature = 0
    engine = nil
    precondition(weakEngine == nil, "Swift model wrapper is still retained")

    try probe.session.appendTokens(tokens: tokens)
    try control.appendTokens(tokens: tokens)
    let position = probe.session.position()
    // Different content and length detect a destructive reset-and-replay of live KV.
    let queryTokens: [UInt32] = [1, 0, 1]
    let states = try probe.hiddenStates(tokens: queryTokens)
    precondition(states.tokenCount == 3 && states.embeddingDim == 32)
    precondition(probe.session.position() == position)
    do {
      _ = try probe.hiddenStates(tokens: [9999])
      fatalError("Invalid token was accepted")
    } catch Cera.FfiError.InvalidToken {}
    precondition(probe.session.position() == position)
    let copied = states.get(token: 0)
    let original = states.data.get(index: 0)
    copied.set(index: 0, value: original + 1)
    precondition(states.data.get(index: 0) == original)
    let actual = try probe.session.generate(opts: options).tokens
    let expected = try control.generate(opts: options).tokens
    precondition(actual == expected && actual.count == 2)
    let bits = states.data.toFloatArray().map { $0.bitPattern }
    precondition(states.data.toFloatArray().allSatisfy { $0.isFinite })
    let result: [String: Any] = [
      "tokens": tokens, "query_tokens": queryTokens, "hidden_bits": bits,
      "generated": actual, "position": position,
    ]
    print(
      String(
        data: try JSONSerialization.data(withJSONObject: result, options: [.sortedKeys]),
        encoding: .utf8)!)
  }
}
